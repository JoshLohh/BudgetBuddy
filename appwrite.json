{
    "projectId": "683b7b97000aae994ff5",
    "projectName": "BudgetBuddy",
    "settings": {
        "services": {
            "account": true,
            "avatars": true,
            "databases": true,
            "locale": true,
            "health": true,
            "storage": true,
            "teams": true,
            "users": true,
            "sites": true,
            "functions": true,
            "graphql": true,
            "messaging": true
        },
        "auth": {
            "methods": {
                "jwt": true,
                "phone": true,
                "invites": true,
                "anonymous": true,
                "email-otp": true,
                "magic-url": true,
                "email-password": true
            },
            "security": {
                "duration": 31536000,
                "limit": 0,
                "sessionsLimit": 10,
                "passwordHistory": 0,
                "passwordDictionary": false,
                "personalDataCheck": false,
                "sessionAlerts": false,
                "mockNumbers": []
            }
        }
    },
    "functions": [
        {
            "$id": "685d1db2003731b3c3d6",
            "execute": [
                "any"
            ],
            "name": "CreateUserProfile",
            "enabled": true,
            "logging": true,
            "runtime": "node-18.0",
            "scopes": [
                "users.read",
                "databases.read",
                "databases.write",
                "collections.read",
                "collections.write",
                "attributes.read",
                "attributes.write",
                "indexes.read",
                "indexes.write",
                "documents.read",
                "documents.write"
            ],
            "events": [
                "users.create"
            ],
            "schedule": "",
            "timeout": 15,
            "entrypoint": "src/main.js",
            "commands": "npm install",
            "specification": "s-0.5vcpu-512mb",
            "path": "functions/CreateUserProfile"
        },
        {
            "$id": "685d26ee003371a58959",
            "execute": [
                "any"
            ],
            "name": "create-user",
            "enabled": true,
            "logging": true,
            "runtime": "bun-1.1",
            "scopes": [
                "users.read",
                "databases.read",
                "databases.write",
                "collections.read",
                "collections.write",
                "attributes.read",
                "attributes.write",
                "indexes.read",
                "indexes.write",
                "documents.read",
                "documents.write"
            ],
            "events": [
                "users.*.create"
            ],
            "schedule": "",
            "timeout": 15,
            "entrypoint": "src/main.ts",
            "commands": "bun install",
            "specification": "s-0.5vcpu-512mb",
            "path": "functions/create-user"
        }
    ],
    "databases": [
        {
            "$id": "684bd404003542c8a2ac",
            "name": "budget_buddy_db",
            "enabled": true
        }
    ],
    "collections": [
        {
            "$id": "684bd42e0009b1e585e7",
            "$permissions": [
                "create(\"users\")",
                "read(\"users\")",
                "update(\"users\")",
                "delete(\"users\")"
            ],
            "databaseId": "684bd404003542c8a2ac",
            "name": "groups",
            "enabled": true,
            "documentSecurity": true,
            "attributes": [
                {
                    "key": "title",
                    "type": "string",
                    "required": true,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "members",
                    "type": "string",
                    "required": false,
                    "array": true,
                    "size": 500,
                    "default": null
                },
                {
                    "key": "createdBy",
                    "type": "string",
                    "required": true,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "description",
                    "type": "string",
                    "required": false,
                    "array": false,
                    "size": 500,
                    "default": null
                }
            ],
            "indexes": []
        },
        {
            "$id": "684bd5ef0030ac0cc7f7",
            "$permissions": [],
            "databaseId": "684bd404003542c8a2ac",
            "name": "expenses",
            "enabled": true,
            "documentSecurity": false,
            "attributes": [
                {
                    "key": "groupId",
                    "type": "string",
                    "required": true,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "title",
                    "type": "string",
                    "required": true,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "amount",
                    "type": "double",
                    "required": true,
                    "array": false,
                    "min": 0,
                    "max": 1.7976931348623157e+308,
                    "default": null
                },
                {
                    "key": "paidBy",
                    "type": "string",
                    "required": true,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "splitBetween",
                    "type": "string",
                    "required": false,
                    "array": true,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "splitType",
                    "type": "string",
                    "required": false,
                    "array": false,
                    "size": 255,
                    "default": "equal"
                },
                {
                    "key": "customSplit",
                    "type": "string",
                    "required": false,
                    "array": false,
                    "size": 1000,
                    "default": null
                }
            ],
            "indexes": []
        },
        {
            "$id": "684bd8a5003d02f9b84d",
            "$permissions": [],
            "databaseId": "684bd404003542c8a2ac",
            "name": "settlements",
            "enabled": true,
            "documentSecurity": false,
            "attributes": [
                {
                    "key": "groupId",
                    "type": "string",
                    "required": true,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "from",
                    "type": "string",
                    "required": true,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "to",
                    "type": "string",
                    "required": true,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "amount",
                    "type": "double",
                    "required": true,
                    "array": false,
                    "min": 0,
                    "max": 1.7976931348623157e+308,
                    "default": null
                },
                {
                    "key": "date",
                    "type": "datetime",
                    "required": true,
                    "array": false,
                    "format": "",
                    "default": null
                }
            ],
            "indexes": []
        },
        {
            "$id": "685cf12e003976d23b39",
            "$permissions": [
                "create(\"any\")",
                "read(\"any\")",
                "update(\"any\")",
                "delete(\"any\")"
            ],
            "databaseId": "684bd404003542c8a2ac",
            "name": "users",
            "enabled": true,
            "documentSecurity": true,
            "attributes": [
                {
                    "key": "userId",
                    "type": "string",
                    "required": true,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "username",
                    "type": "string",
                    "required": true,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "email",
                    "type": "string",
                    "required": true,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "avatar",
                    "type": "string",
                    "required": false,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "bio",
                    "type": "string",
                    "required": false,
                    "array": false,
                    "size": 1000,
                    "default": null
                }
            ],
            "indexes": []
        }
    ]
}