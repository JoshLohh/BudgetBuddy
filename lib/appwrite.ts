import { Client, Account, ID, Avatars } from 'react-native-appwrite';


export const client = new Client()
    .setEndpoint('https://fra.cloud.appwrite.io/v1')
    .setProject("683b7b97000aae994ff5")
    .setPlatform('dev.buddies.BudgetBuddy');

export const account = new Account(client)
export const avatars = new Avatars(client)