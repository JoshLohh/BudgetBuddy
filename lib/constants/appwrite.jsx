import { Client, Account, ID, Avatars } from 'react-native-appwrite';


export const client = new Client
    .setProject("6838ad0d003d2a0754a0")
    .setPlatform('dev.Buddies.budgetbuddy');

export const account = new Account(client)
export const avatars = new Avatars(client)