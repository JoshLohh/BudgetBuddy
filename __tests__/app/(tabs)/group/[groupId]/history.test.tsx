import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react-native';
import GroupHistoryPage from '@/app/(tabs)/group/[groupId]/history';  // adjust path to your file
import { databases } from '@/lib/appwrite';
import { useRouter } from 'expo-router';
import { within } from '@testing-library/react-native';
import type { Mock } from 'jest-mock';

// --- Mocks ---
jest.mock('@/lib/appwrite', () => ({
  databases: {
    listDocuments: jest.fn(),
    getDocument: jest.fn(),
  },
}));
let actualBackMock: jest.Mock<void, []>;

jest.mock('expo-router', () => {
    const actual = jest.requireActual('expo-router');
    actualBackMock = jest.fn();
    const React = require('react');
    return {
      ...actual,
      useRouter: () => ({ back: jest.fn() }),
      useLocalSearchParams: () => ({ groupId: 'testgroupid' }),
      // This makes useFocusEffect run its callback on mount like useEffect!
      useFocusEffect: (callback: any) => React.useEffect(callback, []),
      // useFocusEffect: jest.fn((callback) => React.useEffect(callback, [])),
    };
  });

jest.mock('@expo/vector-icons', () => ({
  Ionicons: () => null,
}));

// --- Test Data ---
const expensesSample = [
  {
    $id: 'e1',
    amount: 12,
    paidBy: 'u1',
    splitBetween: ['u1', 'u2'],
    splitType: 'equal',
    customSplit: null,
    description: 'Lunch',
    groupId: 'testgroupid',
    category: 'Food',
    $createdAt: '2025-05-10T10:30:00Z',
  }
];
const settlementsSample = [
  {
    $id: 's1',
    groupId: 'testgroupid',
    from: 'u2',
    to: 'u1',
    amount: 12,
    $createdAt: '2025-05-12T12:00:00Z',
  }
];
const userProfiles: Record<string, { $id: string; username: string; avatar: string | null }> = {
    u1: { $id: 'u1', username: 'Alice', avatar: 'avatar1.png' },
    u2: { $id: 'u2', username: 'Bob', avatar: null },
  };

beforeEach(() => {
  jest.clearAllMocks();
  (databases.listDocuments as jest.Mock).mockImplementation(() => Promise.resolve({ documents: [] }));
});

it('shows loading initially and then list state', async () => {
  (databases.listDocuments as jest.Mock)
    .mockResolvedValueOnce({ documents: expensesSample })
    .mockResolvedValueOnce({ documents: settlementsSample })
    .mockResolvedValueOnce({ documents: expensesSample })
    .mockResolvedValueOnce({ documents: settlementsSample });

  (databases.getDocument as jest.Mock).mockImplementation(
    (_db, _col, id: string) =>
      Promise.resolve(userProfiles[id] || { $id: id, username: '(unknown)', avatar: null })
  );

  const { getByText, findByTestId } = render(<GroupHistoryPage />);

  expect(getByText(/Loading/i)).toBeTruthy();

  // Find expense activity container by testID (assuming expense $id is 'e1')
  const expenseItem = await findByTestId('activity-e1');
  const { getByText: getByTextInExpense } = within(expenseItem);
  expect(getByTextInExpense(/Alice/i)).toBeTruthy();
  expect(getByTextInExpense(/paid/i)).toBeTruthy();

  // Find settlement activity container by testID (assuming settlement $id is 's1')
  const settlementItem = await findByTestId('activity-s1');
  const { getByText: getByTextInSettlement } = within(settlementItem);
  expect(getByTextInSettlement(/Bob/i)).toBeTruthy();
  expect(getByTextInSettlement(/settled up/i)).toBeTruthy();
});

it('shows "No activity yet" when there is no data', async () => {
    (databases.listDocuments as jest.Mock)
  .mockImplementationOnce(() => Promise.resolve({ documents: [] }))
  .mockImplementationOnce(() => Promise.resolve({ documents: [] }));
    const { findByText } = render(<GroupHistoryPage />);
    expect(await findByText(/No activity yet/)).toBeTruthy();
});

it('displays unknown if user fetch fails', async () => {
    // Mock enough calls for both useEffect & useFocusEffect
    (databases.listDocuments as jest.Mock)
      .mockResolvedValueOnce({ documents: expensesSample })   // expenses
      .mockResolvedValueOnce({ documents: [] })               // settlements
      .mockResolvedValueOnce({ documents: expensesSample })   // expenses (focus)
      .mockResolvedValueOnce({ documents: [] });              // settlements (focus)
  
    // All getDocument calls will fail to fetch
    (databases.getDocument as jest.Mock).mockRejectedValue(new Error('not found'));
  
    const { findByTestId } = render(<GroupHistoryPage />);
    // Find the expense activity container for the activity row
    const expenseItem = await findByTestId('activity-e1');
    const { getByText } = within(expenseItem);
  
    expect(getByText(/\(unknown\)/)).toBeTruthy();
    expect(getByText(/paid/i)).toBeTruthy();
  });

it('toggles sort order via control', async () => {
  (databases.listDocuments as jest.Mock)
    .mockResolvedValueOnce({ documents: expensesSample })
    .mockResolvedValueOnce({ documents: settlementsSample });
  (databases.getDocument as jest.Mock).mockImplementation((_db, _col, id) =>
    Promise.resolve(userProfiles[id])
  );
  const { getByText } = render(<GroupHistoryPage />);
  const sortBtn = getByText(/Newest/i);
  fireEvent.press(sortBtn);
  expect(getByText(/Oldest/i)).toBeTruthy();
});

/*it('calls router.back() when back is pressed', () => {
    const { getByTestId } = render(<GroupHistoryPage />);
    fireEvent.press(getByTestId('back-button'));
    expect(actualBackMock).toHaveBeenCalled();
});*/

it('renders activity log page correctly (snapshot)', async () => {
    // Mock data so that activity items render as expected
    (databases.listDocuments as jest.Mock)
      .mockResolvedValueOnce({ documents: expensesSample })
      .mockResolvedValueOnce({ documents: settlementsSample })
      .mockResolvedValueOnce({ documents: expensesSample })
      .mockResolvedValueOnce({ documents: settlementsSample });
  
    (databases.getDocument as jest.Mock).mockImplementation(
      (_db: any, _col: any, id: string) =>
        Promise.resolve(userProfiles[id] || { $id: id, username: '(unknown)', avatar: null })
    );
  
    const { toJSON, findByTestId } = render(<GroupHistoryPage />);
    // Wait for at least one activity to render (pick the first expense)
    await findByTestId('activity-e1');
    // Snapshot the tree for regression testing (after load & activity shown)
    expect(toJSON()).toMatchSnapshot();
  });
  

// ...add snapshot tests and more edge cases as needed
