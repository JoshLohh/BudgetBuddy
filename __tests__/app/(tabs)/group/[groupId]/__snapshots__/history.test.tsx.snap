// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders activity log page correctly (snapshot) 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#fff",
      },
      {
        "flex": 1,
        "padding": 16,
      },
    ]
  }
  testID="themed-view"
>
  <View
    style={
      {
        "height": 30,
        "width": "100%",
      }
    }
  />
  <View
    accessibilityLabel="Back"
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="back-button"
  />
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 12,
        "marginTop": 10,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          undefined,
          {
            "fontSize": 28,
            "fontWeight": "bold",
            "lineHeight": 32,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          {
            "marginBottom": 12,
            "marginTop": 10,
          },
        ]
      }
    >
      Activity Log
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#f0f0f0",
          "borderRadius": 8,
          "flexDirection": "row",
          "opacity": 1,
          "paddingHorizontal": 10,
          "paddingVertical": 4,
        }
      }
    >
      <Text
        style={
          {
            "color": "#1976d2",
            "fontWeight": "bold",
            "marginLeft": 4,
          }
        }
      >
        Newest
      </Text>
    </View>
  </View>
  <RCTScrollView>
    <View>
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#eafbe7",
            "borderRadius": 10,
            "flexDirection": "row",
            "marginBottom": 10,
            "minHeight": 54,
            "padding": 12,
          }
        }
        testID="activity-s1"
      >
        <View
          style={
            {
              "flex": 1,
              "minWidth": 0,
            }
          }
        >
          <Text
            style={
              [
                {
                  "color": "#11181C",
                },
                {
                  "fontSize": 16,
                  "lineHeight": 24,
                },
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                {
                  "color": "black",
                  "fontSize": 15,
                  "fontWeight": "bold",
                },
              ]
            }
          >
            Bob
          </Text>
          <Text
            numberOfLines={3}
            style={
              [
                {
                  "color": "#11181C",
                },
                {
                  "fontSize": 16,
                  "lineHeight": 24,
                },
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                {
                  "color": "#333",
                  "flexWrap": "wrap",
                  "fontSize": 14,
                  "marginTop": 2,
                },
              ]
            }
          >
            settled up with
             
            <Text
              style={
                [
                  {
                    "color": "#11181C",
                  },
                  {
                    "fontSize": 16,
                    "lineHeight": 24,
                  },
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  {
                    "color": "black",
                    "fontWeight": "bold",
                  },
                ]
              }
            >
              Alice
            </Text>
             
            for
             
            <Text
              style={
                {
                  "color": "#1e88e5",
                  "fontWeight": "bold",
                }
              }
            >
              $
              12.00
            </Text>
          </Text>
          <Text
            style={
              [
                {
                  "color": "#11181C",
                },
                {
                  "fontSize": 16,
                  "lineHeight": 24,
                },
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                {
                  "color": "#777",
                  "fontSize": 12,
                  "marginTop": 2,
                },
              ]
            }
          >
            May 12, 2025
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#f7f7f7",
            "borderRadius": 10,
            "flexDirection": "row",
            "marginBottom": 10,
            "minHeight": 54,
            "padding": 12,
          }
        }
        testID="activity-e1"
      >
        <Image
          source={
            {
              "uri": "avatar1.png",
            }
          }
          style={
            {
              "backgroundColor": "#fff",
              "borderRadius": 16,
              "height": 32,
              "marginRight": 10,
              "width": 32,
            }
          }
        />
        <View
          style={
            {
              "flex": 1,
              "minWidth": 0,
            }
          }
        >
          <Text
            style={
              [
                {
                  "color": "#11181C",
                },
                {
                  "fontSize": 16,
                  "lineHeight": 24,
                },
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                {
                  "color": "black",
                  "fontSize": 15,
                  "fontWeight": "bold",
                },
              ]
            }
          >
            Alice
          </Text>
          <Text
            numberOfLines={3}
            style={
              [
                {
                  "color": "#11181C",
                },
                {
                  "fontSize": 16,
                  "lineHeight": 24,
                },
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                {
                  "color": "#333",
                  "flexWrap": "wrap",
                  "fontSize": 14,
                  "marginTop": 2,
                },
              ]
            }
          >
            paid
             
            <Text
              style={
                {
                  "color": "#1e88e5",
                  "fontWeight": "bold",
                }
              }
            >
              $
              12.00
            </Text>
             
            for "
            Lunch
            "
          </Text>
          <Text
            style={
              [
                {
                  "color": "#11181C",
                },
                {
                  "fontSize": 16,
                  "lineHeight": 24,
                },
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                {
                  "color": "#777",
                  "fontSize": 12,
                  "marginTop": 2,
                },
              ]
            }
          >
            May 10, 2025
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "height": 80,
            "width": "100%",
          }
        }
      />
    </View>
  </RCTScrollView>
</View>
`;
